Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,175 @@
+  }
+
+  func diagnosticReport(for snapshot: DocumentSnapshot, buildSettings: SwiftCompileCommand?, useCache: Bool = false)
+    async throws -> RelatedFullDocumentDiagnosticReport
+  {
+    if useCache, let report = report(for: snapshot.id, buildSettings: buildSettings) {
+      return report
+    }
+    guard let buildSettings = buildSettings else {
+      logger.log(
+        "Producing syntactic diagnostics from the built-in swift-syntax because we don't have buildSettings"
+      )
+      // If we don't have build settings,
+      // sourcekitd won't be able to give us accurate semantic diagnostics.
+      // Fall back to providing syntactic diagnostics from the built-in
+      // swift-syntax. That's the best we can do for now.
+      let report = try await requestFallbackReport(with: snapshot)
+      setReport(for: snapshot.id, buildSettings: nil, report: report)
+      return report
+    }
+    guard !buildSettings.isFallback else {
+      logger.log(
+        "Producing syntactic diagnostics from the built-in swift-syntax because we have fallback arguments"
+      )
+      // If we only have fallback build settings,
+      // sourcekitd won't be able to give us accurate semantic diagnostics.
+      // Fall back to providing syntactic diagnostics from the built-in
+      // swift-syntax. That's the best we can do for now.
+      let report = try await requestFallbackReport(with: snapshot)
+      setReport(for: snapshot.id, buildSettings: buildSettings, report: report)
+      return report
+    }
+    let report = try await requestReport(with: snapshot, compilerArgs: buildSettings.compilerArgs)
+    setReport(for: snapshot.id, buildSettings: buildSettings, report: report)
+    return report
+  }
+}
+
+private extension DiagnosticReportManager {
+  func requestReport(with snapshot: DocumentSnapshot, compilerArgs: [String]) async throws
+    -> LanguageServerProtocol.RelatedFullDocumentDiagnosticReport
+  {
+    try Task.checkCancellation()
+
+    let keys = self.keys
+
+    let skreq = SKDRequestDictionary(sourcekitd: self.sourcekitd)
+    skreq[keys.request] = requests.diagnostics
+    skreq[keys.sourcefile] = snapshot.uri.pseudoPath
+
+    // FIXME: SourceKit should probably cache this for us.
+    skreq[keys.compilerargs] = compilerArgs
+
+    let dict = try await self.sourcekitd.send(skreq, fileContents: snapshot.text)
+
+    try Task.checkCancellation()
+    guard (try? documentManager.latestSnapshot(snapshot.uri).id) == snapshot.id else {
+      // Check that the document wasn't modified while we were getting diagnostics. This could happen because we are
+      // calling `fullDocumentDiagnosticReport` from `publishDiagnosticsIfNeeded` outside of `messageHandlingQueue`
+      // and thus a concurrent edit is possible while we are waiting for the sourcekitd request to return a result.
+      throw ResponseError.unknown("Document was modified while loading document")
```

Comment:
Probably not your typo but mine when I originally wrote this, but might as well fix it.

```suggestion
      throw ResponseError.unknown("Document was modified while loading diagnostics")
```